File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int maxSoFar = 0;
        int maxProfit = 0;
        for (int i = 0; i < n; i++) {
            int maxSoFar = 0;
            for (int j = i; j < n; j++) {
                maxSoFar = max(maxSoFar, prices[j]);
            }
            maxProfit = max(maxProfit, maxSoFar - prices[i]);
        }
        return maxProfit;
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        max_area = 0
        stack = [] # (idx, height)
        for i, h in enumerate(heights):
            start = i
            while stack and stack[-1][1] > h:
                index, height = stack.pop()
                max_area = max(max_area, height * (i - index))
                start = index
            stack.append([start, h])
        n = len(heights)
        while stack:
            index, height = stack.pop()
            max_area = max(max_area, height * (n - index))
        return max_area
        